src/http/output/metadata/AllowAcceptMetadataWriter.ts

```diff
@@ -0,0 +1,99 @@
+import type { HttpResponse } from '../../../server/HttpResponse';
+import { MethodNotAllowedHttpError } from '../../../util/errors/MethodNotAllowedHttpError';
+import { NotFoundHttpError } from '../../../util/errors/NotFoundHttpError';
+import { UnsupportedMediaTypeHttpError } from '../../../util/errors/UnsupportedMediaTypeHttpError';
+import { addHeader } from '../../../util/HeaderUtil';
+import { isContainerPath } from '../../../util/PathUtil';
+import { LDP, PIM, RDF, SOLID_ERROR } from '../../../util/Vocabularies';
+import type { RepresentationMetadata } from '../../representation/RepresentationMetadata';
+import { MetadataWriter } from './MetadataWriter';
+
+/**
+ * Generates Allow, Accept-Patch, Accept-Post, and Accept-Put headers.
+ * The resulting values depend on the choses input methods and types.
+ * The input metadata also gets used to remove methods from that list
+ * if they are not valid in the given situation.
+ */
+export class AllowAcceptMetadataWriter extends MetadataWriter {
+  private readonly supportedMethods: string[];
+  private readonly acceptTypes: { patch: string[]; post: string[]; put: string[] };
+
+  public constructor(supportedMethods: string[], acceptTypes: { patch?: string[]; post?: string[]; put?: string[] }) {
+    super();
+    this.supportedMethods = supportedMethods;
+    this.acceptTypes = { patch: [], post: [], put: [], ...acceptTypes };
+  }
+
+  public async handle(input: { response: HttpResponse; metadata: RepresentationMetadata }): Promise<void> {
+    const { response, metadata } = input;
+
+    // We initially make the assumption that the target resource exists
```

Comment:
```suggestion
    // The XXX and YYY headers vary depending on whether a resource exists,
    // because…
    // We initially make the assumption that the target resource exists
```